File: python/parser.py
from code_generator import code_generator
from lex import lexer

class parser:
    def __init__(self, lex_str, cg_str):
        self.lex = lexer(lex_str)
        self.cg = code_generator(cg_str)
        self.scope_global = []
        self.scope_local = []

    def local_check(self, name):
        for e in self.scope_local:
            if e[0] == name:
                return True
        return False

    def global_check(self, name):
        for e in self.scope_global:
            if e[0] == name:
                print("RETURNING TRUE")
                return True
        print("RETURNING False")
        return False

    def local_params(self):
        c = 0
        for l in self.scope_local:
            if l[1] == False:
                c += 1
        ####print(c)
        return c

    def var_offset(self, name):
        i = 0
        li = 0
        for e in self.scope_local:
            if e[0] == name:
                if e[1] == True:
                    return (i+2)*self.cg.stack
                else:
                    return -1 * (li + 1)*self.cg.stack
            if e[1]:
                i += 1
            else:
                li += 1

    def var_offset_str(self, offset):
        if offset < 0:
            ####print (str(offset))
            return  str(offset)
        elif offset > 0:
            ####print("+"+str(offset))
            return "+"+str(offset)
        else:
            return ""

    def check_token(self, token, expected):
        if token not in expected:
            #print("Expected : ", expected, " Found: ", token)
            return False
        else:
            return True

    # Main function where we are getting the source file.
    def parse_source(self):
        self.std_fun()
        self.program()

    # Just adding the global functions to the scope
    def std_fun(self):
        self.scope_global.append(("####printf", True))
        self.scope_global.append(("malloc", True))
        self.scope_global.append(("calloc", True))
        self.scope_global.append(("free", True))
        self.scope_global.append(("atoi", True))
        self.scope_global.append(("isaplha", True))
        self.scope_global.append(("isdigit", True))
        self.scope_global.append(("isalnum", True))
        self.scope_global.append(("strcpy", True))
        self.scope_global.append(("strlen", True))
        self.scope_global.append(("strcpy", True))
        self.scope_global.append(("strchr", True))
        self.scope_global.append(("strdup", True))
        self.scope_global.append(("fopen", True))
        self.scope_global.append(("fclose", True))
        self.scope_global.append(("fgetc", True))
        self.scope_global.append(("ungetc", True))
        self.scope_global.append(("feof", True))
        self.scope_global.append(("fputs", True))
        self.scope_global.append(("f####printf", True))
        self.scope_global.append(("puts", True))


    def program(self):
        self.cg.write_code_to_file(".intel_syntax noprefix")
        self.lex.next()
        while self.lex.get_token() != None:
            name = self.lex.next()
            current = self.lex.next()
            if current == "(":
                self.decf(name)
            elif current == "=":
                pass

    def decf(self, name):
        ###print("Declare Function called on ", name)
        # Clearing local variables from previous fun
        self.scope_local.clear()
        # Adding syntax
        self.cg.write_code_to_file(".section .text")
        # Adding current function
        self.scope_global.append((name, True))
        #Since we have already passed ( now passing till )
        while self.lex.get_token() != ")":
            datatype = self.lex.next()
            if datatype == ")":
                break
            l_name = self.lex.next()
            self.scope_local.append((l_name, True))
            comma = self.lex.next()
        self.check_token(self.lex.get_token(), {")"})
        tag = self.cg.form_label()
        self.cg.write_label(tag)
        # Cheking for open brackets
        self.check_token(self.lex.next(), {"{"})
        # Parsing inner statements
        self.statements(self.lex.get_token())
        # Adding epilogue
        self.cg.write_function_end()
        # Adding function prologue
        self.cg.write_function_begin(name, self.local_params(), tag)
        pass

    def statements(self, token):
        ###print("Statements")
        if token == '{':
            self.lex.next()
            while self.lex.get_token() != '}':
                self.statements(self.lex.get_token())
            self.check_token(self.lex.get_token(), {"}"})
            self.lex.next()
            ####print("END")
        elif token == "int" or token == "char":
            ####print("INFO:::Found INT/ CHAR")
            name = self.lex.next()
            self.scope_local.append((name, False))
            equal = self.lex.next()
            if equal == "=":
                self.expr(self.lex.next())
                offset = self.var_offset(name)
                self.cg.write_code_to_file("\tmov [ebp"+ self.var_offset_str(offset)+"], eax")
            self.check_token(self.lex.get_token(), ";")
            self.lex.next()
        elif token == "if":
            self.branch(token)
        elif token == "while":
            self.loop(token)
        elif token  == "return":
            self.expr(self.lex.next())
            self.check_token(self.lex.get_token(), ";")
            self.lex.next()
        else:
            self.expr(token)
            self.check_token(self.lex.get_token(), ";")
            self.lex.next()
        pass

    def expr(self, token):
        ###print("EXPR Called")
        self.expr1(token)

    def expr1(self, token):
        ###print("EXPR1 Called")
        self.expr2(token)
        if self.lex.get_token() == '=':
            name = token
            expr = self.lex.next()
            self.expr2(expr)
            offset = self.var_offset(name)
            self.cg.write_code_to_file("\tmov [ebp" + self.var_offset_str(offset) + "], eax")

    def expr2(self, token):
        ###print("EXPR2 Called")
        self.expr3(token)
        operator = self.lex.get_token()
        if operator == "||" or  operator == "&&":
            label = self.cg.form_label()
            self.cg.write_code_to_file("\tcmp eax, 0")
            if operator == "||":
                self.cg.write_code_to_file("\tjnz "+label)
            else:
                self.cg.write_code_to_file("\tjz "+label)
            self.expr3(self.lex.next())
            self.cg.write_label(label)

    def expr3(self, token):
        ###print("EXP3 Called")
        self.expr4(token)
        operator = self.lex.get_token()
        if operator == '<=' or operator == '>=' or operator == '<' or operator == '>' or operator == '!=' or operator == '==':
            self.cg.write_code_to_file("\tpush eax")
            self.expr4(self.lex.next())
            self.cg.write_code_to_file("\tmov ebx, eax")
            self.cg.write_code_to_file("\tpop eax")
            instruction = self.cg.instructions(operator)
            self.cg.write_code_to_file("\tcmp eax, ebx")
            self.cg.write_code_to_file("\t"+instruction+" al")
            self.cg.write_code_to_file("\tmovzx eax, al")


    def expr4(self, token):
        ###print("EXPR4 Called")
        self.expr5(token)
        operator = self.lex.get_token()
        while operator == "+" or operator == "-":
            self.cg.write_code_to_file("\tpush eax")
            self.expr5(self.lex.next())
            self.cg.write_code_to_file("\tmov ebx, eax")
            self.cg.write_code_to_file("\tpop eax")
            instruction = self.cg.instructions(operator)
            self.cg.write_code_to_file("\t"+instruction+" eax, ebx")
            operator = self.lex.get_token()

    def expr5(self, token):
        ###print("EXPR5 Called")
        self.expr6(token)
        operator = self.lex.get_token()
        while operator == "*" or operator == "/":
            self.cg.write_code_to_file("\tpush eax")
            self.expr6(self.lex.next())
            self.cg.write_code_to_file("\tmov ebx, eax")
            self.cg.write_code_to_file("\tpop eax")
            instruction = self.cg.instructions(operator)
            self.cg.write_code_to_file("\t" + instruction + " ebx")
            operator = self.lex.get_token()

    def expr6(self, token):
        ###print("EXPR6 Called")
        if token == "-":
            self.expr6(self.lex.next())
            self.cg.write_code_to_file("\tneg eax")
        else:
            self.unary(token)
            operator = self.lex.get_token()
            if operator == "++" or operator == "--":
                instruction = self.cg.instructions(operator)
                self.cg.write_code_to_file("\t"+instruction+" eax")
                offset = self.var_offset(token)
                self.cg.write_code_to_file("\tmov [ebp" + self.var_offset_str(offset) + "], eax")
                self.lex.next()

    def unary(self, token):
        ###print("UNARY Called")
        name = token
        type = self.lex.token_type
        #print("UNNNAR ",type)
        if type == 2:
            self.cg.write_code_to_file("\tmov eax, "+self.lex.get_token())
        elif type == 4:
            label = self.cg.form_label()
            self.cg.write_code_to_file("\tmov eax, offset "+label)
            self.cg.write_code_to_file(".section .data")
            self.cg.write_code_to_file(label+": .asciz "+ self.lex.get_token())
            self.cg.write_code_to_file(".section .text")
        elif type == 1:
            if self.local_check(name):
                offset = self.var_offset(name)
                self.cg.write_code_to_file("\tmov eax, [ebp" + self.var_offset_str(offset) + "]")
            elif self.global_check(name):
                print("WR")
                self.cg.write_code_to_file("\tlea eax, ["+name+"]")
        curr = self.lex.next()
        if curr == "(":
            self.fun_call(name, self.lex.get_token())

    def fun_call(self, name, token):
        #print("FUN_CALL Called " + name + "  "+ token)
        self.cg.write_code_to_file("\tpush eax")
        self.lex.next()
        s = self.cg.form_label()
        ####print("Start label", s)
        e = self.cg.form_label()
        ####print("end label", e)
        i = 0
        if self.lex.get_token() != ')':
            #print("IF")
            self.cg.write_code_to_file("\tjmp "+s)
            current = self.cg.form_label()
            self.cg.write_label(current)
            #print("BEFORE "+self.lex.get_token())
            self.expr(self.lex.get_token())
            #print("BEFORE2 "+self.lex.get_token())
            self.cg.write_code_to_file("\tpush eax")
            i += 1
            self.cg.write_code_to_file("\tjmp "+e)
            prev = current
            while self.lex.get_token() == ",":
                #print("LPPPP")
                current = self.cg.form_label()
                self.cg.write_label(current)
                self.expr(self.lex.next())
                self.cg.write_code_to_file("\tpush eax")
                i += 1
                self.cg.write_code_to_file("\tjmp "+prev)
                prev = current
            ####print("Start label", s)
            self.cg.write_label(s)
            self.cg.write_code_to_file("\tjmp "+prev)
            self.cg.write_label(e)
        self.cg.write_code_to_file("\tcall [esp+"+str(i*self.cg.stack)+"]")
        self.cg.write_code_to_file("\tadd esp, "+ str((i+1)*self.cg.stack))
        self.check_token(self.lex.get_token(), {")"})
        #print(self.lex.next())
        #print("FUN_CALL End")
        pass

    def branch(self, token):
        ###print("IF Called")
        self.check_token(token, {"if"})
        self.check_token(self.lex.next(), {"("})
        self.expr(self.lex.next())
        e = self.cg.form_label()
        self.cg.write_code_to_file("\tcmp eax, 0")
        self.cg.write_code_to_file("\tjz "+e)
        self.check_token(self.lex.get_token(), {")"})
        i = self.cg.form_label()
        self.cg.write_label(i)
        self.statements(self.lex.next())
        end = self.cg.form_label()
        self.cg.write_code_to_file("jmp "+end)
        self.cg.write_label(e)
        if self.lex.get_token() == "else":
            self.lex.next()
            self.statements(self.lex.get_token())
        #end = self.cg.form_label()
        self.cg.write_label(end)


    def loop(self, token):
        ###print("WHILE Called")
        self.check_token(token, {"while"})
        open = self.lex.next()
        self.check_token(open, {"("})
        e = self.cg.form_label()
        c = self.cg.form_label()
        self.cg.write_label(c)
        self.expr(self.lex.next())
        self.cg.write_code_to_file("\tcmp eax, 0")
        self.cg.write_code_to_file("\tjz "+e)
        self.check_token(self.lex.get_token(), ")")
        b = self.cg.form_label()
        self.cg.write_label(b)
        self.statements(self.lex.next())
        self.cg.write_code_to_file("\tjmp "+c)
        self.cg.write_label(e)

File: go_lang/compiler/parser.go
package compiler

import (
	"fmt"
	"strings"
)

type keyMap struct {
	name  string
	value bool
}

type Parser struct {
	cg     CodeGenerator
	lex    Lex
	global []keyMap
	local  []keyMap
}

func (p Parser) initialize(fileString string, outputFile string) {
	p.lex = Lex{}
	p.cg = CodeGenerator{}
	p.lex.initialize(fileString)
	p.cg.initialize(outputFile)
}

func (p Parser) localCheck(name string) bool {
	for _, element := range p.local {
		if element.name == name {
			return true
		}
	}
	return false
}

func (p Parser) globalCheck(name string) bool {
	for _, element := range p.global {
		if element.name == name {
			return true
		}
	}
	return false
}

func (p Parser) localParams() int {
	var c = 0
	for _, l := range p.local {
		if l.value == false {
			c++
		}
	}
	return c
}

func (p Parser) varOffset(name string) int {
	var offset = 0
	var i = 0
	var li = 0
	for _, element := range p.local {
		if element.name == name {
			if element.value == true {
				return (i + 2) * p.cg.stack
			} else {
				return -1 * (li + 1) * p.cg.stack
			}
		}
		if element.value == true {
			i++
		} else {
			li++
		}
	}
	return offset
}

func (p Parser) varOffsetString(offset int) string {
	if offset < 0 {
		return string(offset)
	} else if offset > 0 {
		return "+" + string(offset)
	} else {
		return ""
	}
	return ""
}

func (p Parser) checkToken(token string, expected []string) bool {
	for _, element := range expected {
		if token == element {
			return true
		}
	}
	fmt.Println("Expected " + strings.Join(expected, ",") + " Found" + token)
	return false
}

func (p Parser) parseSource() {
	p.stdFunctions()
	p.program()
}

func (p Parser) stdFunctions() {
	p.global = append(p.global, keyMap{"printf", true})
}

func (p Parser) program() {
	p.cg.write_code_to_file(".intel_syntax noprefix")
	p.lex.next()
	for p.lex.getToken() != "" {
		var name = p.lex.next()
		var current = p.lex.next()
		if current == "(" {
			p.decf(name)
		} else if current == "=" {
			return
		}
	}
}

func (p Parser) decf(name string) {
	fmt.Println("Declare Function called on" + name)
	p.local = nil
	p.cg.write_code_to_file(".section .text")
	p.global = append(p.global, keyMap{name, true})
	for p.lex.getToken() != ")" {
		var datatype = p.lex.next()
		if datatype == ")" {
			break
		}
		var lName = p.lex.next()
		p.local = append(p.local, keyMap{lName, true})
		_ = p.lex.next()
	}
	p.checkToken(p.lex.getToken(), []string{")"})
	var tag = p.cg.formLabel()
	p.cg.writeLabel(tag)
	p.checkToken(p.lex.next(), []string{"{"})
	p.statements(p.lex.getToken())
	p.cg.functionEnd()
	p.cg.functionBegin(name, p.localParams(), tag)
}

func (p Parser) statements(token string) {
	fmt.Println("Statements")
	if token == "{" {
		p.lex.next()
		for p.lex.getToken() != "}" {
			p.statements(p.lex.getToken())
		}
		p.checkToken(p.lex.getToken(), []string{"}"})
		p.lex.next()
		fmt.Println("End")
	} else if token == "int" || token == "char" {
		var name = p.lex.next()
		p.local = append(p.local, keyMap{name, false})
		var equal = p.lex.next()
		if equal == "=" {
			p.expr(p.lex.next())
			var offset = p.varOffset(name)
			p.cg.write_code_to_file("\tmov [ebp" + p.varOffsetString(offset) + "], eax")
			p.checkToken(p.lex.getToken(), []string{";"})
			p.lex.next()
		}
	} else if token == "if" {
		p.branch(token)
	} else if token == "while" {
		p.loop(token)
	} else if token == "return" {
		p.expr(p.lex.next())
		p.checkToken(p.lex.getToken(), []string{";"})
		p.lex.next()
	} else {
		p.expr(token)
		p.checkToken(p.lex.getToken(), []string{";"})
		p.lex.next()
	}
}

func (p Parser) expr(token string) {
	fmt.Println("Expr")
	p.expr(token)
}

func (p Parser) expr1(token string) {
	fmt.Println("Expr 1")
	p.expr2(token)
	if p.lex.getToken() == "=" {
		var name = token
		var expr = p.lex.next()
		p.expr2(expr)
		var offset = p.varOffset(name)
		p.cg.write_code_to_file("\tmov [ebp" + p.varOffsetString(offset) + "], eax")
	}
}

func (p Parser) expr2(token string) {
	fmt.Println("Expr 2")
	p.expr3(token)
	var operator = p.lex.getToken()
	if operator == "||" || operator == "&&" {
		var label = p.cg.formLabel()
		p.cg.write_code_to_file("\t cmp eax, 0")
		if operator == "||" {
			p.cg.write_code_to_file("\tjnz" + label)
		} else {
			p.cg.write_code_to_file("\tjz" + label)
		}
		p.expr3(p.lex.next())
		p.cg.writeLabel(label)
	}
}

func (p Parser) expr3(token string) {
	fmt.Println("Expr 3")
	p.expr4(token)
	var operator = p.lex.getToken()
	if operator == "<=" || operator == ">=" || operator == "<" || operator == ">" || operator == "!=" || operator == "==" {
		p.cg.write_code_to_file("\tpush eax")
		p.expr4(p.lex.next())
		p.cg.write_code_to_file("\tmov ebx, eax")
		p.cg.write_code_to_file("\tpop eax")
		var instruction = p.cg.instructions(operator)
		p.cg.write_code_to_file("\tcmp eax, ebx")
		p.cg.write_code_to_file("\t" + instruction + "al")
		p.cg.write_code_to_file("\tmovzx eax, al")
	}
}

func (p Parser) expr4(token string) {
	fmt.Println("Expr 4")
	p.expr5(token)
	var operator = p.lex.getToken()
	for operator == "+" || operator == "-" {
		p.cg.write_code_to_file("\tpush eax")
		p.expr5(p.lex.next())
		p.cg.write_code_to_file("\tmov ebx, eax")
		p.cg.write_code_to_file("\tpop eax")
		var instruction = p.cg.instructions(operator)
		p.cg.write_code_to_file("\t" + instruction + " eax, ebx")
		operator = p.lex.getToken()
	}
}

func (p Parser) expr5(token string) {
	fmt.Println("Expr 5")
	p.expr6(token)
	var operator = p.lex.getToken()
	for operator == "*" || operator == "/" {
		p.cg.write_code_to_file("\tpush eax")
		p.expr6(p.lex.next())
		p.cg.write_code_to_file("\tmov ebx, eax")
		p.cg.write_code_to_file("\tpop eax")
		var instructions = p.cg.instructions(operator)
		p.cg.write_code_to_file("\t" + instructions + " ebx")
		operator = p.lex.getToken()
	}
}

func (p Parser) expr6(token string) {
	fmt.Println("Expr 6")
	if token == "-" {
		p.expr6(p.lex.next())
		p.cg.write_code_to_file("\tneg eax")
	} else {
		p.unary(token)
		var operator = p.lex.getToken()
		if operator == "++" || operator == "--" {
			var instruction = p.cg.instructions(operator)
			p.cg.write_code_to_file("\t" + instruction + " eax")
			var offset = p.varOffset(token)
			p.cg.write_code_to_file("\tmov [ebp" + p.varOffsetString(offset) + "], eax")
			p.lex.next()
		}
	}

}

func (p Parser) unary(token string) {
	fmt.Println("Unary Called")
	var name = token
	var tokenType = p.lex.getType()
	switch tokenType {
	case 2:
		p.cg.write_code_to_file("\tmov eax, " + p.lex.getToken())
	case 4:
		var label = p.cg.formLabel()
		p.cg.write_code_to_file("\tmov eax, offset " + label)
		p.cg.write_code_to_file(".section .data")
		p.cg.write_code_to_file(label + ": .asciz " + p.lex.getToken())
		p.cg.write_code_to_file(".section .text")
	case 1:
		if p.localCheck(name) {
			var offset = p.varOffset(name)
			p.cg.write_code_to_file("\tmov eax, [ebp" + p.varOffsetString(offset) + "]")
		} else if p.globalCheck(name) {
			p.cg.write_code_to_file("\tlea eax, [" + name + "]")
		}
		var current = p.lex.next()
		if current == "(" {
			p.funCall(name, p.lex.getToken())
		}

	}

}

func (p Parser) funCall(name string, token string) {
	fmt.Println("Function Called")
	p.cg.writeLabel("\tpush eax")
	p.lex.next()
	var s = p.cg.formLabel()
	fmt.Println("Start Label", s)
	var e = p.cg.formLabel()
	fmt.Println("End label", e)
	var i = 0
	if p.lex.getToken() != ")" {
		p.cg.write_code_to_file("\tjmp" + s)
		var current = p.cg.formLabel()
		p.cg.writeLabel(current)
		p.expr(p.lex.getToken())
		p.cg.write_code_to_file("\tpush eax")
		i++
		p.cg.write_code_to_file("\tjmp " + e)
		var prev = current
		for p.lex.getToken() == "," {
			current = p.cg.formLabel()
			p.cg.writeLabel(current)
			p.expr(p.lex.next())
			p.cg.write_code_to_file("\tpush eax")
			i++
			p.cg.write_code_to_file("\tjmp " + prev)
			prev = current
		}
		fmt.Println("Start label" + s)
		p.cg.writeLabel(s)
		p.cg.write_code_to_file("\tjmp " + prev)
		p.cg.writeLabel(e)
	}
	p.cg.write_code_to_file("\tcall [esp+" + string(i*p.cg.stack) + "]")
	p.cg.write_code_to_file("\tadd esp, " + string((i+1)*p.cg.stack))
	p.checkToken(p.lex.getToken(), []string{")"})
	p.lex.next()
}

func (p Parser) branch(token string) {
	fmt.Println("If Called")
	p.checkToken(token, []string{"if"})
	p.checkToken(p.lex.next(), []string{"("})
	p.expr(p.lex.next())
	var e = p.cg.formLabel()
	p.cg.write_code_to_file("\tcmp eax, 0")
	p.cg.write_code_to_file("\tjz " + e)
	p.checkToken(p.lex.getToken(), []string{")"})
	var i = p.cg.formLabel()
	p.cg.writeLabel(i)
	p.statements(p.lex.next())
	var end = p.cg.formLabel()
	p.cg.write_code_to_file("jmp " + end)
	p.cg.writeLabel(e)
	if p.lex.getToken() == "else" {
		p.lex.next()
		p.statements(p.lex.getToken())
	}
	p.cg.writeLabel(end)
}

func (p Parser) loop(token string) {
	fmt.Println("While Called")
	p.checkToken(token, []string{"while"})
	var open = p.lex.next()
	p.checkToken(open, []string{"("})
	var e = p.cg.formLabel()
	var c = p.cg.formLabel()
	p.cg.writeLabel(c)
	p.expr(p.lex.next())
	p.cg.write_code_to_file("\tcmp eax, 0")
	p.cg.write_code_to_file("\tjz " + e)
	p.checkToken(p.lex.getToken(), []string{")"})
	var b = p.cg.formLabel()
	p.cg.writeLabel(b)
	p.statements(p.lex.next())
	p.cg.write_code_to_file("\tjmp " + c)
	p.cg.writeLabel(e)
}
